theano/gpuarray/tests/test_dnn.py

```diff
@@ -2298,3 +2298,221 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+    img_dims = (10, 256, 256, 3)  # images are usually NHWC
+    img = np.random.randint(low=0, high=256, size=img_dims)
+    # Convert from NHWC to NCHW
+    img = np.transpose(img, axes=(0, 3, 1, 2)).astype(theano.config.floatX)
+    scale_height = 0.25
+    scale_width = 0.75
+
+    # Transformation matrix
+    theta = [[-1, 0, 0],
+             [0, -1, 0]]
+
+    transform = np.asarray(img_dims[0] * [theta], dtype=theano.config.floatX)
+
+    # Create symbolic variables for inputs and transformations
+    t_img = T.tensor4('img')
+    t_theta = T.tensor3('theta')
+
+    st_dnn = dnn.dnn_spatialtf(t_img, t_theta, scale_height=scale_height,
+                               scale_width=scale_width)
+    st_dnn_func = theano.function([t_img, t_theta], st_dnn)
+    # Check if function graph contains the spatial transformer's grid and sampler Ops
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGrid)
+                for node in st_dnn_func.maker.fgraph.toposort()])
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerSampler)
+                for node in st_dnn_func.maker.fgraph.toposort()])
+
+    img_out_gpu = st_dnn_func(img, transform)
+    img_out = np.asarray(img_out_gpu)
+
+    # Setup CPU Op
+    st_cpu = spatialtf_cpu(t_theta, t_img, scale_height, scale_width, 'nearest')
+    st_cpu_func = theano.function([t_theta, t_img], [st_cpu], mode=mode_without_gpu)
+    res, = st_cpu_func(transform, img)
+
+    utt.assert_allclose(img_out, res, rtol=1e-2, atol=1e-2)
+
+
+def test_dnn_spatialtf_grad():
+    if not dnn.dnn_available(test_ctx_name):
+        raise SkipTest(dnn.dnn_available.msg)
+
+    utt.seed_rng()
+
+    # Generate random set of RGB images with 256x256 resolution (pixel values in [0, 255])
+    img_dims = (1, 256, 256, 3)  # images are usually NHWC
+    img = np.random.randint(low=0, high=256, size=img_dims)
+    # Convert from NHWC to NCHW
+    img = np.transpose(img, axes=(0, 3, 1, 2)).astype(theano.config.floatX)
+
+    # Transformation matrix
+    transform = [[-1, 0, 0],
+                 [0, -1, 0]]
+
+    theta = np.asarray(img_dims[0] * [transform], dtype=theano.config.floatX)
+
+    out_shp = (img.shape[0], img.shape[1], int(img.shape[2] * 0.25),
+               int(img.shape[3] * 0.75))
+    dy = -1 + 2 * np.random.randn(*out_shp).astype(theano.config.floatX)
+
+    t_img = T.tensor4('img')
+    t_theta = T.tensor3('theta')
+    t_dy = T.tensor4('dy')
+
+    op = dnn.dnn_spatialtf(t_img, t_theta, scale_height=0.25, scale_width=0.75)
+
+    grad_i, grad_t = T.grad(None, wrt=[t_img, t_theta], known_grads={op: t_dy})
+    grad_fn = theano.function([t_img, t_theta, t_dy], [grad_i, grad_t])
+    grad_fn(img, theta, dy)
+
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGradI)
+                for node in grad_fn.maker.fgraph.toposort()])
+
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGradT)
+                for node in grad_fn.maker.fgraph.toposort()])
```

Comment:
I think all this code is too complicated if the goal is just to check if generated function contains the expected ops !
Also:
1) for gradient test, we don't need to mimic the usual images format, we can just directly generate an NCHW image.
2) To iterate over function graph nodes, you can use `fn.maker.fgraph.apply_nodes` instead of `.toposort()`, it will be faster.

Why not just something like that:
```python
    theta = T.tensor3()
    inputs = T.tensor4()
    output = dnn.dnn_spatialtf(inputs, theta, 0.25, 0.75)
    output_mean = T.mean(output)
    mean_gi = T.grad(output_mean, [inputs])
    mean_gt = T.grad(output_mean, [theta])
    f_gi = theano.function([inputs, theta], mean_gi)
    f_gt = theano.function([inputs, theta], mean_gt)
    assert any(isinstance(node.op, dnn.GpuDnnTransformerGradI) for node in f_gi.maker.fgraph.apply_nodes)
    assert any(isinstance(node.op, dnn.GpuDnnTransformerGradT) for node in f_gt.maker.fgraph.apply_nodes)
    inputs_shape = (1, 3, 10, 10)
    inputs_val = np.random.random(inputs_shape).astype(theano.config.floatX)
    theta_val = np.random.random((inputs_shape[0], 2, 3)).astype(theano.config.floatX)
    # Just check that everything computes.
    f_gi(inputs_val, theta_val)
    f_gt(inputs_val, theta_val)```